operator/pkg/controllers/etcd/etcd.go

```diff
@@ -0,0 +1,346 @@
+	objName := etcdServiceNameFor(controlPlane.ClusterName())
+	service, err := e.getService(ctx, namespacedName(controlPlane.NamespaceName(), objName))
+	if err != nil {
+		if errors.IsNotFound(err) {
+			if err := e.createService(ctx, controlPlane); err != nil {
+				return "", fmt.Errorf("creating etcd service object for cluster %v, %w", controlPlane.ClusterName(), err)
+			}
+			return objName, nil
+		}
+		return "", fmt.Errorf("getting service from api server, %w", err)
+	}
+	return service.Name, nil
+}
+
+func (e *Provider) createService(ctx context.Context, controlPlane *v1alpha1.ControlPlane) error {
+	svc := &v1.Service{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      etcdServiceNameFor(controlPlane.ClusterName()),
+			Namespace: controlPlane.NamespaceName(),
+			Labels:    etcdLabelFor(controlPlane.ClusterName()),
+			OwnerReferences: []metav1.OwnerReference{{
+				APIVersion: controlPlane.APIVersion,
+				Name:       controlPlane.Name,
+				Kind:       controlPlane.Kind,
+				UID:        controlPlane.UID,
+			}},
+		},
+		Spec: v1.ServiceSpec{
+			Ports: []v1.ServicePort{{
+				Port: 2380,
+				Name: etcdServerPortNameFor(controlPlane.ClusterName()),
+			}, {
+				Port: 2379,
+				Name: etcdClientPortNameFor(controlPlane.ClusterName()),
+			}},
+			ClusterIP: "None",
+			Selector:  etcdLabelFor(controlPlane.ClusterName()),
+		},
+	}
+	if err := e.kubeClient.Create(ctx, svc); err != nil {
+		return fmt.Errorf("creating service %s/%s, %w", svc.Namespace, svc.Name, err)
+	}
+	zap.S().Infof("[%s] created service %s", controlPlane.ClusterName(),
+		etcdServiceNameFor(controlPlane.ClusterName()))
+	return nil
+}
+
+func (e *Provider) getService(ctx context.Context, name types.NamespacedName) (*v1.Service, error) {
+	result := &v1.Service{}
+	if err := e.kubeClient.Get(ctx, name, result); err != nil {
+		return nil, fmt.Errorf("getting service %v, %w", name, err)
+	}
+	return result, nil
+}
+
+func (e *Provider) createETCDCerts(ctx context.Context, controlPlane *v1alpha1.ControlPlane) error {
+	// get CA key and cert
+	caCert, caKey, err := e.createCertAndKeyIfNotFound(ctx, controlPlane, etcdRootCACertConfig(controlPlane))
+	if err != nil {
+		return fmt.Errorf("creating root CA for cluster %v, %w", controlPlane.ClusterName(), err)
+	}
+	// generate etcd server, peer, healthcheck and etcdAPIClient certs
+	for _, certConfig := range certListFor(controlPlane, caCert, caKey) {
+		if _, _, err := e.createCertAndKeyIfNotFound(ctx, controlPlane, certConfig); err != nil {
+			return fmt.Errorf("creating certs and key name %s, %w,", certConfig.name, err)
+		}
+	}
+	return nil
+}
+
+// createCertAndKeyIfNotFound checks with management server for a root CA secret, if it exists return the cert and key, else
+// it will generate a new root ca and store as a secret in management server
+func (e *Provider) createCertAndKeyIfNotFound(ctx context.Context, controlPlane *v1alpha1.ControlPlane, certConfig *CertConfig) (*x509.Certificate, crypto.Signer, error) {
```

Comment:
The word you're looking for is `ensure` which is an idempotent create. Also -- construct a return type struct instead of returning multiple types. Happy to discuss multiple return types deeply with you offline, but I see them as something of an anti-pattern outside of the return pattern.